### ISSUE TITLE
Poor performance with steady_tick

### ISSUE THREAD

**Issue by kC0pter** [console-rs/indicatif#389]
While using the steady tick function with a spinner ProgressBar the whole loop just completely slows down.
The version used is the release candidate: 0.17.0-rc.6

```rust
use std::{
    borrow::Cow,
    thread::sleep,
    time::{Duration, Instant},
};

use indicatif::{ProgressBar, ProgressFinish, ProgressIterator, ProgressStyle};

fn main() {
    let iterations = 1000;
    // Set the array with all the blocksizes to test
    let blocksizes: [usize; 7] = [16, 64, 256, 1024, 4096, 16384, 65536];

    // Set the array with all the durations to save
    let mut elapsed: [Duration; 7] = [Duration::ZERO; 7];

    for (pos, blocksize) in blocksizes.iter().enumerate() {
        // Set up the style for the progressbar
        let sty = ProgressStyle::default_spinner()
            .tick_strings(&[
                "▹▹▹▹▹",
                "▸▹▹▹▹",
                "▹▸▹▹▹",
                "▹▹▸▹▹",
                "▹▹▹▸▹",
                "▹▹▹▹▸",
                "▪▪▪▪▪",
            ])
            .template("{prefix} {pos:>4}/{len:4} Iterations per second: {per_sec} {spinner} {msg}")
            .unwrap();

        // Set up the progress bar and apply the style
        let pb = ProgressBar::new(iterations)
            .with_finish(ProgressFinish::WithMessage(Cow::from("done")));
        pb.set_style(sty);
        pb.enable_steady_tick(Duration::from_millis(120));
        pb.set_prefix(format!("Doing test with Blocksize {:5?}:", blocksize));

        // Iterate for the given number of iterations
        // for _ in (0..iterations) {
        for _ in (0..iterations).progress_with(pb) {
            // pb.inc(1);
            // Take a timestamp for timemeasurement later on
            let now = Instant::now();
            sleep(Duration::from_millis(1));
            // Save the elapsed time for later evaluation
            elapsed[pos] += Instant::now() - now;
        }

        // pb.finish_using_style();
    }
}
```

If the steady_tick gets disabled the spinners are not in order since the iterations are too quick.
On the other hand enabling the steady_tick slows the iterations down between 0.5 and 3 per second which it obviously shouldn't since the iteration only takes 1ms + however long the timestamping takes.

I checked if it had anything to do with using a ProgressIterator but it doesn't.
I suspect a race condition somewhere deep down in the thread synchro but i am not sure.
It worked with the release 0.16.2 (though of course with some modifications to functions)

**Comment by sigmaSd** [comment]
I bisected it, the responsible commit seems to be https://github.com/console-rs/indicatif/commit/ea4604399e50847911f9c36d0558129faefc3cf4

**Comment by djc** [comment]
@sigmaSd I'm not sure how you got to that result, but it didn't make sense to me, and when I retried the bisection I got 7e9b1e05708d2548444f5dafd444ad0234b9942e. And indeed, I found that this refactoring failed to drop the lock in the steady ticker before going to sleep, leading to extreme lock contention. 789c7289b353c3e5898695b44af5a3eae716179d contains the simple fix, and I'll push out rc.7 soon.

(Note that I ran a quick benchmark and it seems that this program in release mode takes 0.10s user time vs 0.18s in 0.16.2 on my machine.)
